assembly/src/assembler/mod.rs

```diff
@@ -329,117 +332,76 @@ struct BodyWrapper {
     epilogue: Vec<Operation>,
 }
 
-// TESTS
+// UTILITY FUNCTIONS
 // ================================================================================================
 
-#[test]
-fn nested_block_works() {
-    use crate::{ModuleAst, NamedModuleAst};
-
-    let kernel = r#"
-        export.foo
-            add
-        end"#;
+pub fn combine_blocks(mut blocks: Vec<CodeBlock>) -> CodeBlock {
+    // merge consecutive Span blocks.
+    let mut merged_blocks: Vec<CodeBlock> = Vec::with_capacity(blocks.len());
+    // Keep track of all the consecutive Span blocks and are merged together when
+    // there is a discontinuity.
+    let mut contiguous_spans: Vec<CodeBlock> = Vec::new();
 
-    let assembler = Assembler::new().with_kernel(&kernel).unwrap();
+    blocks.drain(0..).for_each(|block| {
+        if block.is_span() {
+            contiguous_spans.push(block);
+        } else {
+            if !contiguous_spans.is_empty() {
+                merged_blocks.push(combine_spans(&mut contiguous_spans));
+            }
+            merged_blocks.push(block);
+        }
+    });
+    if !contiguous_spans.is_empty() {
+        merged_blocks.push(combine_spans(&mut contiguous_spans));
+    }
 
-    // the assembler should have a single kernel proc in its cache
-    assert_eq!(assembler.proc_cache.len(), 1);
+    // build a binary tree of blocks joining them using Join blocks
+    let mut blocks = merged_blocks;
+    while blocks.len() > 1 {
+        let last_block = if blocks.len() % 2 == 0 {
+            None
+        } else {
+            blocks.pop()
+        };
 
-    // fetch the kernel digest and store into a syscall block
-    let syscall = assembler
-        .proc_cache
-        .values()
-        .next()
-        .map(|p| CodeBlock::new_syscall(p.code_root().hash()))
-        .unwrap();
+        let mut grouped_blocks = Vec::new();
+        core::mem::swap(&mut blocks, &mut grouped_blocks);
+        let mut grouped_blocks = group_vector_elements::<CodeBlock, 2>(grouped_blocks);
+        grouped_blocks.drain(0..).for_each(|pair| {
+            blocks.push(CodeBlock::new_join(pair));
+        });
 
-    struct DummyModuleProvider {
-        module: ModuleAst,
+        if let Some(block) = last_block {
+            blocks.push(block);
+        }
     }
 
-    impl ModuleProvider for DummyModuleProvider {
-        fn get_source(&self, _path: &str) -> Option<&str> {
-            None
-        }
+    blocks.remove(0)
```

Comment:
seems we will panic if the provided blocks are empty. Since the function is infallible, should we add a `debug_assert(!blocks.is_empty())` to the beginning of the function? Alternatively, we can perform this remove only of `blocks` is not empty, and return a CodeBlock with a Noop span